sources/pg_replication/helpers.py

```diff
@@ -0,0 +1,757 @@
+    return {
+        "insert": result[0],
+        "update": result[1],
+        "delete": result[2],
+        "truncate": result[3],
+    }
+
+
+def lsn_int_to_hex(lsn: int) -> str:
+    """Convert integer LSN to postgres' hexadecimal representation."""
+    # https://stackoverflow.com/questions/66797767/lsn-external-representation.
+    return f"{lsn >> 32 & 4294967295:X}/{lsn & 4294967295:08X}"
+
+
+def advance_slot(
+    upto_lsn: int,
+    slot_name: str,
+    credentials: ConnectionStringCredentials,
+) -> None:
+    """Advances position in the replication slot.
+
+    Flushes all messages upto (and including) the message with LSN = `upto_lsn`.
+    This function is used as alternative to psycopg2's `send_feedback` method, because
+    the behavior of that method seems odd when used outside of `consume_stream`.
+    """
+    if upto_lsn != 0:
+        cur = _get_conn(credentials).cursor()
+        cur.execute(
+            f"SELECT * FROM pg_replication_slot_advance('{slot_name}', '{lsn_int_to_hex(upto_lsn)}');"
+        )
+        cur.connection.close()
+
+
+def replication_items(
+    credentials: ConnectionStringCredentials,
+    slot_name: str,
+    pub_name: str,
+    include_columns: Optional[Dict[str, Sequence[str]]] = None,
+    columns: Optional[Dict[str, TTableHintTemplate[TAnySchemaColumns]]] = None,
+    target_batch_size: int = 1000,
+    flush_slot: bool = True,
+    write_disposition: TWriteDisposition = "append",  # TODO: remove after https://github.com/dlt-hub/dlt/issues/1031 has been released
+) -> Iterator[Union[TDataItem, DataItemWithMeta]]:
+    """Yields data items from generator.
+
+    Maintains LSN of last consumed message in state to track progress.
+    At start of the run, advances the slot upto last consumed message in previous run.
+    Processes in batches to limit memory usage.
+    """
+    # start where we left off in previous run
+    resource_name = _gen_replication_resource_name(slot_name, pub_name)
+    start_lsn = dlt.current.resource_state(resource_name).get("last_commit_lsn", 0)
+    if flush_slot:
+        advance_slot(start_lsn, slot_name, credentials)
+
+    # continue until last message in replication slot
+    options = {"publication_names": pub_name, "proto_version": "1"}
+    upto_lsn = get_max_lsn(slot_name, options, credentials)
+    if upto_lsn is None:
+        return "Replication slot is empty."
+
+    # generate items in batches
+    while True:
+        gen = ItemGenerator(
+            credentials=credentials,
+            slot_name=slot_name,
+            options=options,
+            upto_lsn=upto_lsn,
+            start_lsn=start_lsn,
+            target_batch_size=target_batch_size,
+            include_columns=include_columns,
+            columns=columns,
+            write_disposition=write_disposition,  # TODO: remove after https://github.com/dlt-hub/dlt/issues/1031 has been released
+        )
+        yield from gen
+        if gen.generated_all:
```

Comment:
I think in this case we need the cleanup code to only execute if the batch has been yielded completely. We can't split at any LSN, it needs to be the LSN of a `Commit` message. This is because transactions need to be processed atomically, and transactions can consist of multiple messages (e.g. 100 inserted records = 100 messages). The code has been set up such that batch boundaries are always `Commit` messages. Replication can also only start (`start_lsn`) at the beginning of a transaction. So if we have a transaction of 100 inserts, and `add_limit(50)` we don't want to persist resource state, because we haven't yielded the entire transaction.